src/shogun/classifier/Perceptron.cpp

```diff
@@ -17,56 +17,36 @@
 
 using namespace shogun;
 
-CPerceptron::CPerceptron()
-: CLinearMachine()
+CPerceptron::CPerceptron() : CIterativeLinearMachine()
 {
-	init();
-}
-
-CPerceptron::CPerceptron(CDotFeatures* traindat, CLabels* trainlab)
-: CLinearMachine()
-{
-	init();
-	set_features(traindat);
-	set_labels(trainlab);
-}
-
-void CPerceptron::init()
-{
-	max_iter = 1000;
+	m_max_iterations = 1000;
 	learn_rate = 0.1;
 	m_initialize_hyperplane = true;
 	SG_ADD(
 	    &m_initialize_hyperplane, "initialize_hyperplane",
 	    "Whether to initialize hyperplane.", MS_AVAILABLE);
-	SG_ADD(&max_iter, "max_iter", "Maximum number of iterations.", MS_AVAILABLE);
 	SG_ADD(&learn_rate, "learn_rate", "Learning rate.", MS_AVAILABLE);
 }
 
 CPerceptron::~CPerceptron()
 {
 }
 
-bool CPerceptron::train_machine(CFeatures* data)
+void CPerceptron::init_model(CFeatures* data)
 {
-	ASSERT(m_labels)
-
 	if (data)
 	{
 		if (!data->has_property(FP_DOT))
 			SG_ERROR("Specified features are not of type CDotFeatures\n")
 		set_features((CDotFeatures*) data);
+		SG_REF(data);
```

Comment:
more logical would be to put the SG_REF after the assignment and to REF m_continue_features

Reply:
actually, you need to do this
```
SG_REF(data)
SG_UNREF(m_continue_features)
m_continue_features = data
```